### DISCUSSION TITLE
Highlight on hover in very large file hangs; 100% CPU and memory grows

### DISCUSSION THREAD

**Discussion by Atemu** [joaotavora/eglot#1391] [Q&A]
I work on Nix a lot. When I open Nixpkgs' `all-packages.nix` (1.4MiB) with the `nixd` server and hover over an instance of `callPackage`, nixd tells emacs to highlight that keyword and then emacs goes spinning. It can be cancelled but that results in highlighting of `callPackage` instances being stuck.

```
[stderr] I[14:15:13.101] 76380: <-- textDocument/hover(149)
[stderr] I[14:15:13.101] 76380: --> call attrset/attrpathInfo(2092)
[stderr] I[14:15:13.101] 76380: <-- textDocument/documentHighlight(150)
[stderr] I[14:15:13.101] 76380: <-- reply(2092)
[stderr] I[14:15:13.101] 76380: --> reply:textDocument/hover(149)
[server-reply] (id:149) Thu Apr 25 14:15:13 2024:
(:id 149 :jsonrpc "2.0" :result
     (:contents nil :range
                (:end
                 (:character 21 :line 4963)
                 :start
                 (:character 10 :line 4963))))
[stderr] I[14:15:13.224] 76380: --> reply:textDocument/documentHighlight(150)
```

Is it attempting to highlight all 15k instances of `callPackage` in the buffer? Can I make it not do that somehow? I only need the highlighting in the visible portion, the rest can be loaded on-demand.

Alternatively, can I turn of highlight on hover in certain files only? Or ideally in very large files past a certain size.  
If you can't disable just highlighting (that appears to be the primary performance limitation), can you disable eglot in certain paths only (or past a certain size)?

**Comment by joaotavora** [comment]
> Is it attempting to highlight all 15k instances of callPackage in the buffer? 

Possibly. Let's confirm that by your experiment after `M-x profiler-start` (choose the "CPU+MEM" profiler), then `M-x profiler-stop` after you're done experimenting, then `M-x profiler-report`, and press `RET` (or maybe `C-u RET`)` to expand a selected interesting subtrees.  Then post the result here.

But perhaps , and very importantly, before doing that, make sure you report your Emacs version, your Eglot version, and desirably everything (and I really do mean everything) requested in https://www.gnu.org/software/emacs/manual/html_node/eglot/Reporting-bugs.html
